### PR TITLE
Pr imx219

### PR DESCRIPTION
**Pull request by caleridas**
imx219 driver with the following caveats:
- only 8bpp modes supported, may be limitation of csi bridge (10bpp modes will just hang)
- v4l2 ctrls still broken on the mxc capture driver, not clear
- added the driver to "cleaned up" new DTS, but this is likely not exactly the way we want it

PR fine for review making the thing runnable end-to-end, but probably needs some more fixes before merging.

To test with the sample capture tool:
./sample-capture -w 3280 -h 2464 -f RGGB

Any other v4l2 capture will work as well, but it critically needs to select correct bpp mode.

### CODE REVIEW

**Comment by ostaebler** [comment]
Regarding b44624557b1f7da40c575cd10aa5a3f393c7f208: Why don't you cherry-pick upstream commits?
I think these two would do the same:
https://source.codeaurora.org/external/imx/linux-imx/commit/?h=imx_5.4.3_2.0.0&id=7a9b109d91cfc6089006378efd515cc287bdef67
https://source.codeaurora.org/external/imx/linux-imx/commit/?h=imx_5.4.3_2.0.0&id=95ce9c28601afc5da0c11792601ad32dd14cdd44

This makes future updates easier.

File: drivers/media/i2c/imx219.c [hunk 1]
```diff
@@ -1274,9 +1274,7 @@ static void imx219_free_controls(struct imx219 *imx219)
 static int imx219_check_hwcfg(struct device *dev)
 {
 	struct fwnode_handle *endpoint;
-	struct v4l2_fwnode_endpoint ep_cfg = {
-		.bus_type = V4L2_MBUS_CSI2_DPHY
-	};
+	struct v4l2_fwnode_endpoint* ep_cfg = NULL;
```
Comment (ostaebler) [inline drivers/media/i2c/imx219.c:1277]:
```suggestion
	struct v4l2_fwnode_endpoint *ep_cfg = NULL;
```

File: drivers/media/i2c/imx219.c [hunk 2]
```diff
@@ -1285,34 +1283,35 @@ static int imx219_check_hwcfg(struct device *dev)
 		return -EINVAL;
 	}
 
-	if (v4l2_fwnode_endpoint_alloc_parse(endpoint, &ep_cfg)) {
+	ep_cfg = v4l2_fwnode_endpoint_alloc_parse(endpoint);
+	if (!ep_cfg) {
```
Comment (ostaebler) [inline drivers/media/i2c/imx219.c:1287]:
```
/**
 * v4l2_fwnode_endpoint_alloc_parse() - parse all fwnode node properties
 ...
 * Return: Pointer to v4l2_fwnode_endpoint if successful, on an error pointer
 * on error.
 */
```
```suggestion
	if (IS_ERR(ep_cfg) {
```

**Comment by caleridas** [comment]
> Regarding [b446245](https://github.com/Auterion/linux-tn-imx/commit/b44624557b1f7da40c575cd10aa5a3f393c7f208): Why don't you cherry-pick upstream commits?
> I think these two would do the same:
> https://source.codeaurora.org/external/imx/linux-imx/commit/?h=imx_5.4.3_2.0.0&id=7a9b109d91cfc6089006378efd515cc287bdef67
> https://source.codeaurora.org/external/imx/linux-imx/commit/?h=imx_5.4.3_2.0.0&id=95ce9c28601afc5da0c11792601ad32dd14cdd44
> 
> This makes future updates easier.

right, will change it -- originally just quickly hacked these trivial functions to get it running

File: drivers/media/i2c/imx219.c [hunk 3]
```diff
@@ -1459,6 +1458,12 @@ static const struct of_device_id imx219_dt_ids[] = {
 };
 MODULE_DEVICE_TABLE(of, imx219_dt_ids);
 
+static const struct i2c_device_id imx219_id[] = {
+	{ "imx219", 0 },
+	{}
+};
+MODULE_DEVICE_TABLE(i2c, imx219_id);
```
Comment (ostaebler) [inline drivers/media/i2c/imx219.c:1465]: This seems unused. Maybe add `.id_table = imx219_id,` on line 1476?

File: arch/arm64/boot/dts/freescale/imx8mm-pico-pi-skynodev2_dev_3.dts [hunk 4]
```diff
@@ -271,3 +269,72 @@
...
+			mipicam_ep: endpoint {
+				remote-endpoint = <&mipi1_sensor_ep>;
+				data-lanes = <1 2>;
+				clock-lanes = <0>;
+				clock-noncontinuous;
+				link-frequencies = /bits/ 64 <456000000>;
+			};
+		};
+
+		imx219_clk: camera-clk {
```
Comment (ostaebler) [inline arch/arm64/boot/dts/freescale/imx8mm-pico-pi-skynodev2_dev_3.dts:334]: Clocks usually are defined under the root node. E.g. `arch/arm64/boot/dts/freescale/imx8mm-axon-pi.dts:102`

File: arch/arm64/boot/dts/freescale/skynodev2_p2.dts [hunk 5]
```diff
@@ -825,3 +825,66 @@
...
+			mipicam_ep: endpoint {
+				remote-endpoint = <&mipi1_sensor_ep>;
+				data-lanes = <1 2>;
+				clock-lanes = <0>;
+				clock-noncontinuous;
+				link-frequencies = /bits/ 64 <456000000>;
+			};
+		};
+
+		imx219_clk: camera-clk {
```
Comment (ostaebler) [inline arch/arm64/boot/dts/freescale/skynodev2_p2.dts:884]: Clocks usually are defined under the root node. E.g. arch/arm64/boot/dts/freescale/imx8mm-axon-pi.dts:102

File: arch/arm64/boot/dts/freescale/skynodev2_p2.dts [hunk 6]
```diff
@@ -825,3 +825,66 @@
...
+
+		csi1_mipi_ep: endpoint2 {
+			remote-endpoint = <&csi1_ep>;
+		};
+	};
+};
+
+/* camera */
+&i2c3 {
+	imx219_mipi: imx219_mipi@10 {
```
Comment (ostaebler) [inline arch/arm64/boot/dts/freescale/skynodev2_p2.dts:863]: Label not used/necessary.

**Comment by ostaebler** [comment]
Squash f6439569f0d4a7b4c80b58955f7cb10e14b4b004 into 6e94933cdb01af0a284a0be096d609376ed81cf8?

**Comment by davids5** [comment]
OK I found the issue. Something is fundamental broken in the it does not get back to a reset state. 

First run with correct mode
```
root@auterionmachine:/root# ./simple-capture-arm64.dat -w 3280 -h 2464 -f RGGB
Supported: RG10 (10-bit Bayer RGRG/GBGB) max: 3280x2464
Supported: RGGB (8-bit Bayer RGRG/GBGB) max: 3280x2464
Selected: RGGB 3280x2464
Capturing 20 images
```

Second run with incorrect param.

```
root@auterionmachine:/root# ./simple-capture-arm64.dat
Supported: RG10 (10-bit Bayer RGRG/GBGB) max: 3280x2464
Supported: RGGB (8-bit Bayer RGRG/GBGB) max: 3280x2464
Selected: RG10 2592x1944
Capturing 20 images
^Cpture 0
```

Third run with correct mode **AFTER** wrong params.
```
root@auterionmachine:/root# ./simple-capture-arm64.dat -w 3280 -h 2464 -f RGGB
Supported: RG10 (10-bit Bayer RGRG/GBGB) max: 3280x2464
Supported: RGGB (8-bit Bayer RGRG/GBGB) max: 3280x2464
Selected: RGGB 3280x2464
Capturing 20 images
Capture 0
```

**Pull request merged**